src/pymor/reductors/spectral_factor.py

```diff
@@ -0,0 +1,93 @@
+# This file is part of the pyMOR project (https://www.pymor.org).
+# Copyright pyMOR developers and contributors. All rights reserved.
+# License: BSD 2-Clause License (https://opensource.org/licenses/BSD-2-Clause)
+
+import numpy as np
+
+from pymor.core.base import BasicObject
+from pymor.models.iosys import LTIModel
+from pymor.parameters.base import Mu
+from pymor.algorithms.lyapunov import solve_cont_lyap_dense
+from pymor.algorithms.to_matrix import to_matrix
+from pymor.algorithms.lyapunov import _chol
+from pymor.operators.numpy import NumpyMatrixOperator
+from pymor.reductors.h2 import IRKAReductor
+from pymor.reductors.bt import BTReductor
+
+class SpectralFactorReductor(BasicObject):
+    r"""
+    Passivity preserving model reduction via spectral factorization.
+
+    See :cite:`BU22`.
+    """
+
+    def __init__(self, fom, mu=None):
+        assert isinstance(fom, LTIModel)
+        if not isinstance(mu, Mu):
+            mu = fom.parameters.parse(mu)
+        assert fom.parameters.assert_compatible(mu)
+        self.fom = fom
+        self.mu = mu
+
+    def reduce(self, r=None):
+        # TODO Use operators directly instead of converting to dense matrix
+        # TODO where possible.
+        E = to_matrix(self.fom.E, format='dense')
+        A = to_matrix(self.fom.A, format='dense')
+        B = to_matrix(self.fom.B, format='dense')
+        C = to_matrix(self.fom.C, format='dense')
+        D = to_matrix(self.fom.D, format='dense')
+
+        # Compute minimal X
+        Z = self.fom.gramian('pr_o_lrcf', mu=self.mu).to_numpy()
+        # TODO Add option to supply X from outside?
+        # TODO Do we really need to compute the full X out of the low-rank factor Z?
+        # TODO However, currently, `solve_pos_ricc_lrcf` uses a dense solver in
+        # TODO the background anyway?
+        X = Z.T@Z
+        
+        # Compute Cholesky-like factorization of W(X)
+        M = _chol(D+D.T).T
+        # TODO Alternatives to taking matrix inverse?
+        L = np.linalg.solve(M.T, C-B.T@X@E)
+        # TODO Currently, relative LTL error too high?
+        LTL = -A.T@X@E - X@A@E
+        relLTLerr = np.linalg.norm(L.T@L-LTL)/np.linalg.norm(LTL)
+        self.logger.info(f'Relative L^T*L error: {relLTLerr:.3e}')
+        LTM = C.T - E.T@X@B
+        relLTMerr = np.linalg.norm(L.T@M-LTM)/np.linalg.norm(LTM)
+        self.logger.info(f'Relative L^T*M error: {relLTMerr:.3e}')
+
+        spectral_factor = LTIModel(self.fom.A, self.fom.B,
+            NumpyMatrixOperator(L, source_id=self.fom.A.range.id),
+            NumpyMatrixOperator(M, source_id=self.fom.B.source.id),
+            self.fom.E)
+        
+        # TODO Allow to set other reductor or reductor options from outside
```

Comment:
It should be possible to set the reductor and/or reductor options used for the spectral factor from the outside. @pmli @sdrave Do you have any preferences how this should be achieved? I see the following two options:

- Allow to pass a lambda function as an argument of `SpectralFactorReductor.reduce(...)` which performs the reduce step for the spectral factor.
- Inheritance (i.e. add `IRKASpectralFactorReductor`, `BTSpectralFactorReductor` etc. reductors  which inherit from a `GenericSpectralFactorReductor` base class), similarily as it is done for balanced truncation in https://github.com/pymor/pymor/blob/main/src/pymor/reductors/bt.py with `GenericBTReductor`, `BTReductor`, `FDBTReductor` etc.

Reply:
I guess inheritance would be too much here. Passing a function could be a good way.

I feel that a "nicer" way would be to pass a reductor class, but to be flexible as much as possible, `SpectralFactorReductor.reduce` would need to accept `spectral_factor_reductor`, `spectral_factor_reductor_opts`, and `spectral_factor_reductor_reduce_opts` (maybe with better names), and call

```python
spectral_factor_reduced = spectral_factor_reductor(
    spectral_factor,
    self.mu,
    **spectral_factor_reductor_opts,
).reduce(r, **spectral_factor_reductor_reduce_opts)
```

This would be somewhat similar to [SOR-IRKA](https://github.com/pymor/pymor/blob/9f3252e624c117f4b0108679c31db6af5025f74b/src/pymor/reductors/sor_irka.py).

Reply:
What would be the advantage of expecting a class (or in fact an arbitrary factory method for something with a `reduce` method) over just expecting a `reduce` function with the right arguments? That might be the case if all relevant reductors are constructed in the same way. The other question would be how to select a default reductor. Or would a reductor always have to be specified. Of course, with the class approach, one could just write the reductor class as the default value.